enhancements/authentication/invalid-certs.md

```diff
@@ -0,0 +1,200 @@
+2. Verifies if above metrics have a value >0
+3. If invalid certificates are detected, the controller sets a `InvalidCertsUpgradeable=false` condition on the `cluster` `KubeAPIServer.operator.openshift.io` resource which will yield the `Upgradeable=false` status on the `kube-apiserver` `config.openshift.io.ClusterOperator` resource.
+
+#### oauth-server
+
+For oauth-server this enhancement proposes the introduction of the following new metric:
+
+- `openshift_auth_external_x509_missing_san_total`: a metric capturing the count of invalid certificates
+presented by an external identity provider.
+
+To detect invalid certificates this enhancement proposes to add a new controller in `cluster-authentication-operator` that:
+
+1. Regularly executes queries against the cluster internal Thanos instance
+2. Verifies if above metrics have a value >0
+3. If invalid certificates are detected, the controller sets a `InvalidCertsUpgradeable=false` condition on the `cluster` `Authentication.operator.openshift.io` resource which will yield the `Upgradeable=false` status on the `authentication` `config.openshift.io.ClusterOperator` resource.
+
+#### cluster-authentication-operator
+
+In case of having invalid custom certificates configured for the external OAuth route
+the existing `RouterCertsDomainValidationController` is extended.
+Formally additional logic is added in `validateRouterCertificates` https://github.com/openshift/cluster-authentication-operator/blob/ff156ab2bdfbdd68b49d76547fea5ec28d9c3639/pkg/controllers/routercerts/controller.go#L129.
+
+Additionally, custom routes controller in `openshift-authentication-operator` will reject invalid certificates in https://github.com/openshift/cluster-authentication-operator/blob/d2f6218a6ab2daccbec43f71a495de1c80529ef6/pkg/controllers/customroute/custom_route_controller.go#L55.
+
+#### library-go
+
+A library-go function will be implemented that takes a `*x509.Certificate` and returns a `bool` value
+indicating if it is an invalid certificate. The function will check whether the given certificate relies
+on the Common Name field rather than SANs.
+
+This function will be used to assert validity of invalid certificates in the external OAuth route case.
+
+The higher level controller which asserts metrics for a given Prometheus client will be provided in libary-go.
+
+### Risks and Mitigations
+
+This enhancement provides a means to protect against broken clusters when upgrading from OpenShift 4.9 to OpenShift 4.10.
+
+## Design Details
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+
+N/A
+
+### Graduation Criteria
+
+N/A
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+The above proposed changes **must** be backported to OpenShift 4.9.
+Otherwise, a detection of invalid certificates will not be possible before an upgrade to OpenShift 4.10.
```

Comment:
I would also be interested in what happens in future releases.

Scenario: a new 4.11 feature triggers calls to a new HTTPS endpoint. The feature, that doesn't exist in 4.10, is active in 4.11 independently of user intervention. The new HTTPS endpoint is served with what you defined as "invalid certificate".

In this scenario, upgrading from 4.10 to 4.11 potentially brings the cluster to an unstable state.

In general terms: for each new HTTPS endpoint we call in 4.y, do we commit on adding an HTTPS-cert check in 4.y-1.z? Do we commit to doing that until a specific date?

For context, [Chrome has dropped support for CN in v58](https://developers.google.com/web/updates/2017/03/chrome-58-deprecations#remove_support_for_commonname_matching_in_certificates) in April 2017.

Reply:
Starting from 4.10 we're not having a go runtime any more that is capable to accept legacy cert based endpoints in the first place.

Reply:
It seemed to me that [library-go's IsHostnameError](https://github.com/openshift/library-go/pull/1275/files#diff-02272a0bc61d7d6f395f4226c37361bad47d9a5acc37f6fc28519e424be29e40R25-R40) was specifically designed for checking invalid certificates under Go v1.17+. What am I missing?

Reply:
No, `IsHostnameError` will only work with <=1.16.